Applications/Spire/Source/BookView/MarketHighlightsTableView.cpp

```diff
@@ -0,0 +1,477 @@
+        auto input_box = new InputBox(m_model);
+        input_box->set_read_only(true);
+        layout()->addWidget(input_box);
+        setFocusProxy(input_box);
+      }
+    }
+  };
+
+  auto make_item(const std::shared_ptr<ExpandableTableModel>& expandable_table,
+      const MarketDatabase& markets, const std::shared_ptr<TableModel>& table,
+      int row, int column) {
+    if(column == MARKET_COLUMN) {
+      auto market_box = make_market_box(
+        make_table_value_model<MarketCode>(table, row, column),
+        expandable_table->m_available_markets[row], markets);
+      auto item = new EditableBox(*market_box);
+      if(row == table->get_row_size() - 1) {
+        match(*item, EmptyMarket());
+      }
+      market_box->get_current()->connect_update_signal(
+        [=] (const auto& market) {
+          if(!market.IsEmpty() && is_match(*item, EmptyMarket())) {
+            unmatch(*item, EmptyMarket());
+          }
+        });
+      return item;
+    } else if(column == LEVEL_COLUMN) {
+      auto settings = setup_level_box();
+      settings.m_current =
+        make_table_value_model<MarketHighlightLevel>(table, row, column);
+      return new EditableBox(*new InputBoxWrapper<EnumBox<MarketHighlightLevel>,
+        EnumBox<MarketHighlightLevel>::Settings>(std::move(settings),
+          make_table_value_model<MarketCode>(table, row, MARKET_COLUMN)));
+    } else if(column == COLOR_COLUMN) {
+      return new EditableBox(*new InputBoxWrapper<HighlightBox,
+        std::shared_ptr<HighlightColorModel>>(
+          make_table_value_model<HighlightColor>(table, row, column),
+          make_table_value_model<MarketCode>(table, row, MARKET_COLUMN)));
+    }
+    throw std::runtime_error("Invalid column.");
+  }
+}
+
+TableView* Spire::make_market_highlights_table_view(
+    std::shared_ptr<MarketHighlightListModel> market_highlights,
+    MarketDatabase markets, QWidget* parent) {
+  auto table = std::make_shared<ExpandableTableModel>(
+    std::make_shared<MarketHighlightListToTableModel>(
+      market_highlights, COLUMN_COUNT, &extract), markets);
+  auto table_builder = EditableTableViewBuilder(table);
+  table_builder.add_header_item(QObject::tr("Market"), QObject::tr("Mkt"),
+    TableHeaderItem::Order::NONE, TableFilter::Filter::NONE);
+  table_builder.add_header_item(QObject::tr("Highlight"), QObject::tr("HL"),
+    TableHeaderItem::Order::UNORDERED, TableFilter::Filter::NONE);
+  table_builder.add_header_item(QObject::tr("Color"), QObject::tr("Clr"),
+    TableHeaderItem::Order::UNORDERED, TableFilter::Filter::NONE);
+  table_builder.set_item_builder(std::bind_front(make_item, std::move(table),
+    markets));
+  table_builder.set_comparator([=] (const AnyRef& left, int left_row,
+      const AnyRef& right, int right_row, int column) {
+    auto& left_market = any_cast<MarketCode>(left);
+    auto& right_market = any_cast<MarketCode>(right);
+    if(left_market.IsEmpty() || right_market.IsEmpty()) {
```

Comment:
This violates the weak ordering property of a comparator. The empty market has to either be the smallest or the largest but it can't be both.